// chqportal/app.js
var setting = require('chqportal').setting;



var class_MenuHighlight = "selected color-blue-500";
var sm = $("#navigation-sidemenu");
var smli = sm.find(".menu > li:not(.divider)");
var smitems = sm.find(".menu > li:not(.divider) > a");
var mc = $(".main-content");


function hashChange(e) {
	if (location.hash == "" || $("[href='" + location.hash + "']").length == 0) {
		location.hash = "#dashboard";
		return;
	}
	if (Responsive.device != "desktop") {
		SideMenu.hide(sm.get(0));
	}
	smli.removeClass(class_MenuHighlight);
	var a = $("[href='" + location.hash + "']");
	var title = a.text();
	setTitle(title);
	a.parent().addClass(class_MenuHighlight)
	mc.scrollTop(0);
}

function init() {
	var md = new Material();
	theme();
	screen();
	window.onhashchange = hashChange;
	hashChange();
	key();
}

function screen() {
	var isFullScreen = setting.getFullScreen();
	if(isFullScreen=="true" || isFullScreen)
	{
		$("#full-screen-switch").attr("checked", true);
	}
}


function theme() {
	var theme = setting.getTheme();
	if (theme == "dark") {
		$("#dark-theme-switch").attr("checked", true);
	}
	setTheme(theme);
}

function key() {
	$(document.body).on("keydown", function(e) {
		var key = String.fromCharCode(e.keyCode);
		if (e.altKey && key >= "1" && key <= "9") {
			var index = parseInt(key, 10) - 1;
			var hash = smli.eq(index).find("a").attr("href");
			if(hash)
			{
				location.hash = hash;
			}
		}
	});
}

function setTitle(str) {
	$(".header .toolbar-label").html(str);
}

function setTheme(theme) {
	Theme.setTheme(theme);
	setting.setTheme(theme);
}

module.exports.init = init;
module.exports.setTheme = setTheme;
module.exports.setTitle = setTitle;
module.exports.theme = theme;
