// chqportal/userSetting.js
'use strict';

var Datastore = require('nedb');
var path = require('path');
var dbPath = path.resolve(__dirname, "db");
var userSettingPath = path.resolve(dbPath, "userSetting.db");
var db = new Datastore({
	filename: userSettingPath
});
var common = require('./common.js');
var defaultSetting = {
	fullScreen: false,
	theme: "light",
	dashboard: [{
		id: "default",
		name: "DEFAULT",
		shortcuts: []
	}]
}


function _set(prop, value, cb) {
	var uptObj = {};
	uptObj[prop] = value;
	db.update({}, {
		$set: uptObj
	}, {}, function(err, numReplaced) {
		if (!err) {
			console.log(numReplaced + " records update:", prop, value);
			if (typeof cb == "function") cb();
		}
	})
}

function _get(prop, cb) {
	db.find({}, function(err, doc) {
		if (typeof cb == "function") {
			if (doc && doc[0]) {
				cb(err, doc[0][prop])
			}
		}
	})
}

module.exports.init = function(cb) {
	if (typeof cb != "function") cb = function() {}
	db.loadDatabase();
	db.count({}, function(err, count) {
		if (!err) {
			if (count == 0) {
				db.insert(defaultSetting, function(err, docs) {
					if (!err) {
						db.find({}, function(err, docs) {
							if (err) cb(err)
							else {
								common.set("userSetting", docs[0])
								cb(null, docs[0])
							}
						})
					} else cb(err);
				})
			} else {
				db.find({}, function(err, docs) {
					if (err) cb(err)
					else {
						common.set("userSetting", docs[0])
						cb(null, docs[0])
					}
				})
			}
		} else cb(err);
	});
}

module.exports.reload = function(cb) {
	if (typeof cb != "function") cb = function() {};
	db.find({}, function(err, docs) {
		if (err) cb(err)
		else {
			common.set("userSetting", docs[0])
			cb(null, docs[0])
		}
	})
}

module.exports.getConnection = function() {
	return db;
}



// function log(str) {
// 	require('./shell.js').copy(str);
// }

module.exports.get = _get;
module.exports.set = _set;
